File: ruby/robot-name/robot_name.rb
class Robot
  attr_reader :name

  def initialize
    generate_name
  end

  def reset
    generate_name
  end

  private

  def generate_name
    @name = ('A'..'Z').to_a.sample(2).join
    @name << (1..10).to_a.sample(3).join
  end
end

File: go/robot-name/robot_name.go
package robotname

import (
	"crypto/rand"
	"math/big"
)

type Robot struct {
	name string
}

func (robot *Robot) Name() string {
	if robot.name == "" {
		robot.name = robot.generateName()
	}
	return robot.name
}

func (robot *Robot) Reset() string {
	robot.name = robot.generateName()
	return robot.name
}

func (robot *Robot) generateName() string {
	bytes := make([]byte, 5)

	for i := 0; i < 3; i++ {
		bytes[i] = byte(secureRandom(26) + 65)
	}
	for i := 2; i < 5; i++ {
		bytes[i] = byte(secureRandom(10) + 48)
	}

	return string(bytes)
}

func secureRandom(max int64) int64 {
	val, _ := rand.Int(rand.Reader, big.NewInt(max))
	return val.Int64()
}
